flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/MailboxReceiver.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.runtime.tasks.mailbox;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+/**
+ * Consumer-facing side of the {@link Mailbox} interface. This is used to dequeue letters. The mailbox returns letters
+ * in the order by which they were enqueued. A mailbox should only be consumed by one thread at a time.
+ */
+public interface MailboxReceiver {
+
+	/**
+	 * Returns <code>true</code> if the mailbox contains mail.
+	 */
+	boolean hasMail();
+
+	/**
+	 * Returns the oldest letter from the mailbox (head of queue) if the mailbox is not empty or
+	 * <code>null</code> otherwise.
+	 *
+	 * @return the oldest letter from the mailbox (head of queue) if the mailbox is not empty or
+	 * <code>null</code> otherwise.
+	 */
+	@Nullable
+	Runnable tryTakeMail();
```

Comment:
`Optional`?

Reply:
I would prefer to keep this as returning `null` because it is a data structure and therefore in line with what JDK queues etc do.

Reply:
Same reasoning should apply to `InputGate` for example and this is the same discussion that we had before. `@Nullable` is unfortunately worthless in our code base :( Since performance doesn't matter here, I would prefer `Optional` far more compared to maintaining a similarity with a Java collections - unless you want to implement their interfaces.

Reply:
Even outside the question about how JDK does it, this class is just a data-structure, at the lowest level of abstraction and I feel like there is not truly any benefit a "forced protection" on this level. I can see benefits further up the stack, where semantics of interfaces start to be less clear about what to expect. Or would you retroactively like to add `Optional` returns on classes like `ArrayList` if you could?

Reply:
Retroactively I would enforce compile time checks for `@Nullable` :) But aside that, I wouldn't add `Optional` to java collections, because they are more generic and there are use cases where performance overhead of `Optional` matter. Here, that's not the case.